Use page list instead of placeholder as fallback

draganescu opened the pull request:
Closes #42563

- [x] Default to a page list block when the navigation block is empty and no menus exist
  - [x] Add test for this
  - [x] Update tests that assume empty nav defaults to placeholder
    - [x] Update `selectClassicMenu` to pull the block toolbar instead of placeholder
  - [x] Update the front end to default to a page list block when the navigation block is empty and no menus exist
- [x] Show "empty placeholder" when deleting the selected menu from the block
  - [ ] Add test for this
- [x] Default to the most recently created menu if multiple menus exist
  - [x] Add test for this
- [x] Default to the only menu if only one menu exists
  - [x] Add test for this
- [x] 🐛 deleting the fallback page list block results in a menu being created, i think this is correct **for now**
- [x] 🐛 inserting two blocks when there are no menus (fallback on page list) automatically creates a new menu
- [ ] 🐛 responsive navigation blocks don't allow the fallback page list to be selected via mouse click, , opened #43227

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This makes the navigation block default to a page list in the editor when loading an empty navigation block (one that has no uncontrolled blocks nor any ref).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We do this so that the editor respects the fallback of the front end, which is a page list.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

We use a similar technique to the one used in the pattern block which does not add the pattern before it is edited.

## Testing Instructions

1. Use a theme that has a template part with an empty navigation block (e.g. Vivre)
2. Load the template part in the site editor
3. Make sure you delete ALL your navigation menus
4. You should see a page list block for navigation

## Screenshots or screencast <!-- if applicable -->

N/A

getdave commented:
Whatever we do here has to include changes from https://github.com/WordPress/gutenberg/pull/42182 in order to ensure they work together.

Mamaduka commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -559,6 +560,23 @@ function Navigation( {
 		{ open: overlayMenuPreview }
 	);
 
+	useEffect( () => {
+		if ( isFallback ) {
+			__unstableMarkNextChangeAsNotPersistent();
+			replaceInnerBlocks( clientId, [ createBlock( 'core/page-list' ) ] );
+		}
+	}, [ isFallback ] );
+
+	const innerBlocksProps = useInnerBlocksProps( blockProps );
```
Is there any reason for not directly setting `'core/page-list'` as an inner blocks template?

draganescu replied:
It doesn't really work because we need to not register a save and we will always look to see if the innerBlocks have been there from the markup or if they're added programatically as a fallback. UnsavedInnerBlocks will create an undo when adding the template.

Mamaduka replied:
Thanks for checking, @draganescu.

scruffian commented:
Some things I noticed, perhaps not related to this change or necessary to fix in this PR
1. When the fallback is a page list block, I can't select it in the editor
2. When I edit the block so that its using a navigation CPT, the template appears unedited in the editor:
<img width="1020" alt="Screenshot 2022-08-01 at 14 19 49" src="https://user-images.githubusercontent.com/275961/182157815-d3ddc25b-58e0-4c28-b945-a8052108b958.png">

I found this too: https://github.com/WordPress/gutenberg/issues/42856

draganescu commented:
@scruffian the template part seems to pick up the changes for me:

https://user-images.githubusercontent.com/107534/182177319-737034d7-9129-467a-bd6b-e555e9077f1c.mp4

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -547,7 +555,9 @@ function Navigation( {
 			if ( ! ref ) {
-				replaceInnerBlocks( clientId, [] );
+				replaceInnerBlocks( clientId, [
+					createBlock( 'core/navigation-link' ),
```
This needs a comment to explain what is happening and why.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -263,17 +263,24 @@ function Navigation( {
+		}
+
+		if ( navigationMenus?.length === 1 ) {
+			__unstableMarkNextChangeAsNotPersistent();
```
We should explain _why_ we need to do this in a comment.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -263,17 +263,24 @@ function Navigation( {
-		setRef( navigationMenus[ 0 ].id );
-	}, [ navigationMenus ] );
+		if ( ! navigationMenus?.length ) {
+			__unstableMarkNextChangeAsNotPersistent();
```
We should explain _why_ we need to do this in a comment.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -690,6 +700,24 @@ function Navigation( {
 		</InspectorControls>
 	);
 
+	if ( isPlaceholder && CustomPlaceholder ) {
```
We should explain _why_ the placeholder is still here even though generally it's deprecated.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -710,6 +738,20 @@ function Navigation( {
+					<BlockControls>
+						<ToolbarGroup className="wp-block-navigation__toolbar-menu-selector">
+							<NavigationMenuSelector
+								ref={ null }
+								currentMenuId={ null }
```
Can we explain why these are set to `null`? I assume because we don't have a ref to a post at the moment?

draganescu replied:
This control will go away anyway to the block inspector so the code will go away eventually too. It is here only as an escape hatch for now.

getdave reviewed (commented):
Thank you for working on this tricky problem.

https://user-images.githubusercontent.com/444434/182183787-6c706c49-c8a1-4a4f-a65f-a8b8bab8e940.mov

On first pass I wanted to stress test this PR by ignoring the instructions and just coming to it with no expectations.

❌ I opened the Post Editor and added a Nav block and I **ended up in an unresolvable state** with no way to select, create or add items to my menu. I had x4 Navigation Menu posts in the database when I did this.

✅ When I manually deleted all my Navigation Menus and then added a Nav block I was presented with a Page List. 

✅  When I then saved that Navigation and then opened a new Post and added a new Navigation block, the block automatically selected the first Navigation menu.

❌ When I tried to create a new Navigation Menu via the `Create new` menu item nothing happened and I went **back to an unresolvable state**.

----

At this point I'd say there a few side effects and problems that we need to iron out. Given the complexity of the scope of the block's interactions, I often find that adding e2e tests helps me to keep track of avoiding regressions and also exposing edge cases. You might want to explore that.


PS: "unresolvable state" looks like this

<img width="1002" alt="Screenshot 2022-08-01 at 16 32 20" src="https://user-images.githubusercontent.com/444434/182185568-6b20585b-14e6-402c-8044-c927386f80f1.png">

scruffian commented:
If I have
- a navigation block in the header and another in the footer
- no navigation menus
- I set the header to use a classic menu
- the footer also gets updated to use the classic menu, because now I have a CPT

This feel unintuitive, but it also does make sense when you understand why....

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -254,19 +254,30 @@ function Navigation( {
 			return;
 		}
 
+		navigationMenus.sort( ( menuA, menuB ) => {
```
This will need a complementary change to mirror it on the PHP side on the front of the site

draganescu replied:
Done in [552da29](https://github.com/WordPress/gutenberg/pull/42735/commits/552da2917d490439298dafbfaf7ad9ff4bc688ec)

getdave reviewed (commented):
I tested this principally by creating x2 menus in Empty Theme and then switching to TT2.

- Initially it rendered the most recent menu (both Editor and FoS)
- I then manually deleted 1 of the x2 menus.
- I then reloaded. It rendered the single menu (both Editor and FoS)
- I then manually deleted the remaining menu.
- It then rendered the Page List (both Editor and FoS).

This seems good. Will continue testing.

## Missing e2e tests

We'll want/need to update the e2e tests to match the new functionality. I know @talldan is rewriting these in Playwright so that's going to cause some pain.

## Issues

I did notice the following when testing in the Post editor:


### Creating a Menu then deleting leaves empty block on front of site

I created a Menu and saved it. Then I deleted it and when back to the Editor. I saw the "your block is not longer available message" but on the front of the site there was no fallback displayed.

### I created a new menu and saw permissions warning

Update: [I've tried to fix this over here](https://github.com/WordPress/gutenberg/pull/42940).

I deleted a menu and then refreshed the post. Editor showed me the msg to warn about deleted menu but when I click on the message it shows me a permissions warning.

https://user-images.githubusercontent.com/444434/182622513-7269fb77-7389-44f5-9aa0-3317304824b8.mp4

### Empty Navigation renders empty on front end

I created a new menu and left it empty. Then I saved. I switched to the front of the site and I saw an empty menu. This is expected _but_ the `<nav>` element was still rendered. Should it not rendering no block at all?

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -219,12 +219,24 @@ function Navigation( {
 			ref ||
-			! navigationMenus?.length ||
-			navigationMenus?.length > 1
+			! navigationMenus?.length
```
This change will require alterations to the comment above:

> // - there is more than 1 Navigation Post (wp_navigation).

getdave replied:
This effect could become a hook to further simplify the Nav block edit definition.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -219,12 +219,24 @@ function Navigation( {
 			hasUncontrolledInnerBlocks ||
 			isCreatingNavigationMenu ||
 			ref ||
-			! navigationMenus?.length ||
-			navigationMenus?.length > 1
```
This change will require alterations to the comment above:

> // - there is more than 1 Navigation Post (wp_navigation).

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -257,6 +269,17 @@ function Navigation( {
 		hasResolvedNavigationMenus &&
 		! hasUncontrolledInnerBlocks;
 
+	if ( isPlaceholder && ! ref ) {
```
```suggestion
	if ( isFallback && ! ref ) {
```

Do you think we ought now to rename the term "placeholder" to avoid it being confused with `<Placeholder>` components, especially as we will provide one of these components for backwards compat.

draganescu replied:
I did think about this as well but the page list block is a placeholder and we also have the custom placeholder that depends on `isPlaceholder`.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -257,6 +269,17 @@ function Navigation( {
 		! hasUncontrolledInnerBlocks;
 
+	if ( isPlaceholder && ! ref ) {
+		/**
+		 *  this fallback only displays (both in editor and on front)
+		 *  the list of pages block if not menu is available
+		 *  we don't want the fallback to request a save
+		 *  nor to be undoable, hence we mark it non persistent
+		 */
```
```suggestion
		/**
		 *  This fallback displays (both in editor and on front)
		 *  a list of pages only if no menu (user assigned or
		 * automatically picked) is available.
		 *  The fallback should not request a save (entity dirty state)
		 *  nor to be undoable, hence why it is marked as non persistent
		 */
```

Nit: some tweaks.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -687,7 +716,15 @@ function Navigation( {
 		: Placeholder;
 
-	if ( isPlaceholder ) {
+	/**
+	 * The navigation block supported Cusrtom Placeholders.
+	 * Even though the current UX tries as hard as possible to not
+	 * end up in a placeholder state, the block continues to support
+	 * this extensibility point, via a CustomPlaceholder.
+	 * WHen CustomPlaceholder is present it becomes the default fallback
+	 * for an empty navigation block, instead of the default fallbacks.
+	 */
```
```suggestion
	/**
	 * Historically the navigation block has supported custom placeholders.
	 * Even though the current UX tries as hard as possible not to
	 * end up in a placeholder state, the block continues to support
	 * this extensibility point, via a CustomPlaceholder.
	 * When CustomPlaceholder is present it becomes the default fallback
	 * for an empty navigation block, instead of the default fallbacks.
	 */
```

Nit

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -716,7 +753,9 @@ function Navigation( {
 								onSelect={ handleSelectNavigation }
-								onCreateNew={ resetToEmptyBlock }
+								onCreateNew={ () =>
+									createNavigationMenu( '', [] )
```
At some point we should consider extracting a utility function which creates an empty Nav block. It would improve the code readability.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -735,7 +774,7 @@ function Navigation( {
 								<NavigationMenuDeleteControl
 									onDelete={ ( deletedMenuTitle = '' ) => {
-										resetToEmptyBlock();
+										createNavigationMenu( '', [] );
```
Ditto above.

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
 
 /**
- * Finds the first non-empty `wp_navigation` Post.
+ * Finds the most recent published `wp_navigation` Post.
```
```suggestion
 * Finds the most recently published `wp_navigation` Post.
```

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
+		'orderby'        => 'date',
 		'post_status'    => 'publish',
-		'posts_per_page' => 20, // Try the first 20 posts.
+		'posts_per_page' => 1, // Try the first 20 posts.
```
```suggestion
		'posts_per_page' => 1, // get only the most recent
```

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
-			return $navigation_post;
-		}
+	$navigation_post = new WP_Query( $parsed_args );
+	if( count( $navigation_post->posts )  > 0 ) {
```
```suggestion
	if( count( $navigation_post->posts ) > 0 ) {
```

Nit

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
 
-	$navigation_posts = new WP_Query( $parsed_args );
-	foreach ( $navigation_posts->posts as $navigation_post ) {
-		if ( has_blocks( $navigation_post ) ) {
-			return $navigation_post;
-		}
```
Why did we ditch the `has_blocks` check here?

draganescu replied:
If it's published and empty it is still valid :) IF the saving as draft idea works this makes even more sense.

getdave replied:
I just wonder _why_ we would have introduced this check if it served no purpose? 

When I saw it removed without explanation it triggered my spidey senses 😄

draganescu replied:
I think it was that we _always_ wanted to show _something_ but with this new workflow it seems this is not the case. I hope removing it will not cause some weirdness, but since we want to respect editor and front parity of content, if the editor has a published empty menu and no items in it so will the front end, so this condition makes little sense.

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
-		}
+	$navigation_post = new WP_Query( $parsed_args );
+	if( count( $navigation_post->posts )  > 0 ) {
+		return $navigation_post->posts[0];
```
```suggestion
		return $navigation_post->post;
```

This should work

draganescu replied:
Why would it? WP_Query returns an array with one element.

getdave replied:
https://github.com/WordPress/wordpress-develop/blob/67023d0c66267420107bb565f887ae6a95b12666/src/wp-includes/class-wp-query.php#L118-L127

But yes maybe that's not intended for this use case.

draganescu replied:
I feel like `post` was for `the_post()` and I think this clunkiness `posts[0]` makes it clear we're only about one post.

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -325,12 +320,13 @@ function block_core_navigation_get_fallback_blocks() {
 	// Default to a list of Pages.
 
-	$navigation_post = block_core_navigation_get_first_non_empty_navigation();
+	$navigation_post = block_core_navigation_get_most_recent_published_navigation();
```
```suggestion
	$navigation_post = block_core_navigation_get_most_recently_published_navigation();
```

draganescu replied:
Renamed

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
-	// see:
-	// - https://github.com/WordPress/wordpress-develop/blob/ba943e113d3b31b121f77a2d30aebe14b047c69d/src/wp-includes/nav-menu.php#L613-L619.
-	// - https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters.
+function block_core_navigation_get_most_recent_published_navigation() {
```
```suggestion
function block_core_navigation_get_most_recently_published_navigation() {
```

draganescu replied:
Renamed!

getdave commented on packages/block-library/src/navigation/index.php:
```diff
@@ -325,12 +320,13 @@ function block_core_navigation_get_fallback_blocks() {
 	if ( $navigation_post ) {
 		$maybe_fallback = block_core_navigation_filter_out_empty_blocks( parse_blocks( $navigation_post->post_content ) );
 
+
```
```suggestion
```

Not needed?

draganescu replied:
Indeed 👯

getdave reviewed (commented):
Some code stuff I noticed.

jasmussen commented:
Thanks for working on this. Took it for a quick spin and it looks like I need a little guidance on my local setup, and possibly I found a few small glitches. 

![site editor](https://user-images.githubusercontent.com/1204802/183024401-d348c993-bccb-4ddf-8587-1f83d4643549.gif)

It's a bit hard to discern what's going on, and this wasn't as pure a test as it should be since I had a lot of menus to delete. But:

* It looks like new menus are created when inserting fresh, or deleting a menu, as opposed to selecting an existing menu. Perhaps a separate issue?
* Deleting menus didn't appear to work as intended — deleting "Header navigation 3" appeared to create a duplicate of the same name, and then suddenly a new "Header navigation 2".

I'll see if I can get these menus deleted and try again.

scruffian reviewed (approved):
This is looking great. I think we should merge once you have addressed Dave's comments :)

draganescu commented:
Addressed the feedback in [f679f69](https://github.com/WordPress/gutenberg/pull/42735/commits/f679f698345f32590ae66ac7f57d6fff7b5e9ace). Will do a new round of review and fix/add tests

jasmussen commented:
Took it for a spin, long GIF with description after:
![fallback](https://user-images.githubusercontent.com/1204802/183362386-4072b352-0144-4f51-892d-fbd10e861fa3.gif)

Deleted all the menus, deleting goes to the fallback placeholder. After that, I inserted a new nav menu which fell back to the page list. 

Instead of showing a placeholder, I'd still fall back to the "menu selection logic" which would load the first available menu when inserting fresh. But nothing that need block this PR (unless that's trivial to implement). Nice work 👏

A followup challenge is making it then easy to customize the Page List for when you want a custom menu (which can be a rabbit hole unless we're careful) — one thing we likely do need to fix is that right now you can't select the Page List child block unless you use the list view:

![cant select](https://user-images.githubusercontent.com/1204802/183363252-650b7259-f4b0-47df-9e95-5072cfcd832d.gif)

This is the case when the mobile responsive feature is enabled, and it looks like the "responsive container" intercepts the click. How best to fix this, I'm not sure — touching `pointer-events` always feels fragile 🤔

talldan commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -257,6 +270,17 @@ function Navigation( {
+		 *  nor to be undoable, hence we mark it non persistent
+		 */
+		__unstableMarkNextChangeAsNotPersistent();
+		replaceInnerBlocks( clientId, [ createBlock( 'core/page-list' ) ] );
```
It looks like this action is dispatched on render, so have to be a little bit careful about accidentally dispatching it multiple times. 

Using the inner blocks `template` property may also be a viable way to define some default blocks. Though I might be missing some aspects of the requirements here.

draganescu replied:
@talldan is [this](https://github.com/WordPress/gutenberg/pull/42735#discussion_r932981300) a similar idea?

draganescu committed:
add page list fallback, fix tests

scruffian committed:
update comment

scruffian commented:
This mostly looks good. ~The problem I see is that even when I have any wp_navigation post types the fallback is always to Page List. We could deal with that in a different PR though....~ This was a faulty test, sorry.

draganescu committed:
PHP linting

scruffian committed:
Update packages/block-library/src/navigation/index.php

Co-authored-by: Dave Smith <getdavemail@gmail.com>

scruffian committed:
Update packages/block-library/src/navigation/index.php

Co-authored-by: Dave Smith <getdavemail@gmail.com>

scruffian commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -210,19 +204,37 @@ function Navigation( {
+	// - the user is creating a new menu.
+	// - there are no menus to choose from.
+	// This attempts to pick the first menu if there is a single Navigation Post. If more
+	// than 1 exists then no attempt to automatically pick a menu is made.
```
```suggestion
	// than 1 exists then use the most recent.
```

scruffian committed:
Update packages/block-library/src/navigation/edit/index.js

scruffian commented:
LGTM (again!)

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -250,29 +250,24 @@ function block_core_navigation_render_submenu_icon() {
+		'orderby'        => 'date',
 		'post_status'    => 'publish',
-		'posts_per_page' => 20, // Try the first 20 posts.
+		'posts_per_page' => 1, // get only the most recent
```
```suggestion
		'posts_per_page' => 1, // get only the most recent.
```

scruffian committed:
Update packages/block-library/src/navigation/index.php

draganescu commented:
Found a new bug while debugging tests: inserting two blocks when there are no menus (fallback on page list) automatically creates a new menu.

draganescu committed:
updated tests to wait for network idle when a menu is created

draganescu committed:
update tests await for clicks and create navigations to test for view JS

draganescu committed:
fix auto menu creation for multiple empty blocks inserted

draganescu committed:
improve a comment

draganescu commented:
I think the issue in https://github.com/WordPress/gutenberg/issues/43227 blocks the tests from passing because when using interactive if I click on the navigation block the pass is 100%. Sometimes the test seems to be stuck at the `await waitForInserterCloseAndContentFocus();` in the `insertBlock( searchTerm ) {` helper and this did not happen before this PR.

draganescu committed:
fix rebase bad merge of

scruffian committed:
remove the __unstableIsDisabled prop from the navigation block

scruffian commented:
I pushed a commit ([6c67ecd](https://github.com/WordPress/gutenberg/pull/42735/commits/6c67ecd8982dec30116a71b1da91dbb403c54bb2)) which removes the "disabled" state of the block when there's an overlay. I hope this fixes https://github.com/WordPress/gutenberg/issues/43227. I don't think we need this state anymore.

draganescu committed:
fix error where converting a classic menu did not select the newly created block menu. also added a missing await in the test classic menu selecting function

jasmussen commented:
🚀

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -1463,22 +1703,22 @@ Expected mock function not to be called but it was called with: ["POST", "http:/
+				"//button[@aria-checked='false'][contains(., 'First navigation')]"
 			);
-			await exampleNavigationOption.click();
+			theOption.click();
```
@draganescu There's another `await` here. Also noticed this test seems to be skipped now?

draganescu replied:
🤦🏻  I'll make a separate PR to add the await and also check out what's up with the skipping. I think the missing awaits and [6c67ecd](https://github.com/WordPress/gutenberg/pull/42735/commits/6c67ecd8982dec30116a71b1da91dbb403c54bb2)) make it possible to remove the skipping.

getdave commented:
@draganescu I think we need to ensure we add a dev note for this one. The change I wrote about [here](https://github.com/WordPress/gutenberg/pull/42182) which got merged into this PR means that any Themes which include a Page List in their bundled template parts will cause their users to effectively "opt out" of the default fallback experience.

It's actually quite a significant change as many Themes have taken to including Page List in their Nav block inner blocks.

draganescu commented:
### Navigation Block Fallback Behavior in WP 6.1 Dev Note

In WordPress 6.1 the navigation block will have a new fallback behavior.

**What is the fallback behavior?**

When a theme uses a navigation block in a template part , the aim is to hint at where the navigation should visually be located, for the theme's UI to be consistent, usable and good looking. However, themes can't know beforehand what menus the site has, how many pages, what they're called and so on. For this reason, themes include a navigation block in template parts, but without knowing what content they'll show.

The fallback behavior is the small heuristics in the navigation block which tries to determine what the block should display, by default, when a user activates a theme.

**What is the new fall back behavior?**

Starting with WordPress 6.1, theme developers and authors can lean on the following fallback behavior of the block:

If the navigation block has inner blocks, it will honor them and display that. If it is empty however, then it will initialize the fallback behavior.

The fallback behavior (in both the editor and the front of the site) is:

- if there are no block menus or classic menus, the block will display a list of available pages using the Page List block.
- if there are multiple block menus, the navigation block will display the most recently created block menu.

The key changes can be summarised as follows:

1. Improved consistency.
2. Page List as default fallback.
3. Selecting the most recent block menu.

**Consistency:** previously the fallback behavior was not consistent between the front of the site and the editor. If a theme used an empty navigation block it would display a list of pages on the front, and an empty block in the editor. Now the behavior is consistent between both meaning that the editor mirrors what users see on the front.

**Defaulting to page list:** previously, themes which wanted to default to a page list in the editor usually included a page list inner block within the Navigation block. With this update this is no longer necessary, as the block, if empty, will automatically have  consistent front and editor behavior, eventually defaulting to a page list.

**Selecting the most recent block menu:** this part of the fallback behavior is new, in the event a site has multiple block menus, an empty navigation block will display the most recent one.

**Notes**

_Display only._

The fallback behavior only affects what the empty navigation block will display. Unless the user edits the navigation block's default fallback, adding a link, changing a label, converting a page list block to a list of links or selecting another menu, the markup of the template part is not changed.

_Default content is still honored._

There is no change to how navigation blocks with inner blocks from theme markup behave. Themes  still include inner blocks in the markup in the event they want to showcase a specific situation, for instance a small three links menu, pointing to #, with some restriction on length of link labels - this will continue to work, just like before, rendering the uncontrolled inner blocks both on the front and in the editor.

**Props**

Dave Smith (@getdave ) for editing and technical review.
